File: core/src/main/java/pcb/uwu/amount/composite/mechanics/Force.java
package pcb.uwu.amount.composite.mechanics;

import pcb.uwu.amount.quantity.Mass;
import pcb.uwu.core.BigDecimalAmount;
import pcb.uwu.core.CompositeUnitAmount;
import pcb.uwu.core.Magnitude;
import pcb.uwu.core.UnitAmount;
import pcb.uwu.units.composite.mechanics.AccelerationUnit;
import pcb.uwu.units.composite.mechanics.ForceUnit;
import pcb.uwu.units.composite.mechanics.SpeedUnit;
import pcb.uwu.units.quantity.LengthUnit;
import pcb.uwu.units.quantity.MassUnit;
import pcb.uwu.units.quantity.TimeUnit;

import java.math.BigDecimal;
import java.math.MathContext;

public class Force extends CompositeUnitAmount<ForceUnit> {

	// region constructors

	public Force(Number value, ForceUnit unit) {
		super(value, unit);
	}

	public Force(Number value, Magnitude magnitude, ForceUnit unit) {
		super(value, magnitude, unit);
	}

	public Force(String value, ForceUnit unit) {
		super(value, unit);
	}

	public Force(String value, Magnitude magnitude, ForceUnit unit) {
		super(value, magnitude, unit);
	}

	public Force(BigDecimal value, ForceUnit unit) {
		super(value, unit);
	}

	public Force(BigDecimal value, Magnitude magnitude, ForceUnit unit) {
		super(value, magnitude, unit);
	}

	public Force(BigDecimalAmount amount, ForceUnit unit) {
		super(amount, unit);
	}

	public Force(BigDecimalAmount amount, Magnitude magnitude, ForceUnit unit) {
		super(amount, magnitude, unit);
	}

	// endregion

	// region implement UnitAmount

	@Override
	public Force plus(UnitAmount<ForceUnit> other, MathContext mathContext) {
		return new Force(plusAmount(other, mathContext), getUnit());
	}

	@Override
	public Force minus(UnitAmount<ForceUnit> other, MathContext mathContext) {
		return new Force(minusAmount(other, mathContext), getUnit());
	}

	@Override
	public Force multipliedBy(BigDecimal other, MathContext mathContext) {
		return new Force(multipliedByScalar(other, mathContext), getUnit());
	}

	@Override
	public Force dividedBy(BigDecimal other, MathContext mathContext) {
		return new Force(dividedByScalar(other, mathContext), getUnit());
	}

	@Override
	public Force convertTo(ForceUnit unit) {
		return new Force(getAmountIn(unit), unit);
	}

	// endregion

	// region composition

	public Mass dividedBy(Acceleration acceleration, MathContext mathContext) {
		MassUnit massUnit = getUnit().getUnitCounter().findUnit(MassUnit.class);

		return new Mass(super.dividedBy(acceleration, mathContext).getAmount(), massUnit);
	}

	public Acceleration dividedBy(Mass mass, MathContext mathContext) {
		LengthUnit lengthUnit = getUnit().getUnitCounter().findUnit(LengthUnit.class);
		TimeUnit timeUnit = getUnit().getUnitCounter().findUnit(TimeUnit.class);

		AccelerationUnit accelerationUnit = new AccelerationUnit(new SpeedUnit(lengthUnit, timeUnit), timeUnit);

		return new Acceleration(super.dividedBy(mass, mathContext).getAmount(), accelerationUnit);
	}

	// endregion
}

File: dimension/src/main/java/pcb/uwu/amount/derived/mechanics/Force.kt
package pcb.uwu.amount.derived.mechanics

import pcb.uwu.amount.base.Length
import pcb.uwu.amount.base.Mass
import pcb.uwu.amount.derived.geometry.Area
import pcb.uwu.amount.derived.thermodynamics.Energy
import pcb.uwu.amount.derived.thermodynamics.Pressure
import pcb.uwu.core.CompositeUnitAmount
import pcb.uwu.core.Magnitude
import pcb.uwu.core.Magnitude.NATURAL
import pcb.uwu.core.UnitAmount
import pcb.uwu.unit.derived.geometry.AreaUnit
import pcb.uwu.unit.derived.mechanics.AccelerationUnit
import pcb.uwu.unit.derived.mechanics.ForceUnit
import pcb.uwu.unit.derived.termodynamics.EnergyUnit
import pcb.uwu.unit.derived.termodynamics.PressureUnit
import pcb.uwu.util.UnitAmountUtils

open class Force : CompositeUnitAmount<ForceUnit>
{
    @JvmOverloads
    constructor(amount: Number,
                magnitude: Magnitude = NATURAL,
                unit: ForceUnit)
            : super(amount, magnitude, unit)

    @JvmOverloads
    constructor(amount: String,
                magnitude: Magnitude = NATURAL,
                unit: ForceUnit)
            : super(amount, magnitude, unit)

    // region UnitAmount

    @Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
    override fun plus(force: UnitAmount<ForceUnit>) =
        Force(amount = this.amount + (force to this.unit).amount,
              unit = this.unit)

    @Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
    override fun minus(force: UnitAmount<ForceUnit>) =
        Force(amount = this.amount - (force to this.unit).amount,
              unit = this.unit)

    override fun times(scalar: Number) =
        Force(amount = this.amount * scalar,
              unit = this.unit)

    override fun div(scalar: Number) =
        Force(amount = this.amount / scalar,
              unit = this.unit)

    override fun to(unit: ForceUnit) =
        Force(amount = UnitAmountUtils.getAmountIn(unitAmount = this, newUnit = unit),
              unit = unit)

    // endregion

    // region composition

    operator fun div(mass: Mass) =
        Acceleration(amount = this.amount / mass.amount,
                     unit = AccelerationUnit(forceUnit = this.unit,
                                             massUnit = mass.unit))

    open operator fun times(length: Length) =
        Energy(amount = this.amount * length.amount,
               unit = EnergyUnit(forceUnit = this.unit,
                                 lengthUnit = length.unit))

    open operator fun div(pressure: Pressure) =
        Area(amount = this.amount / pressure.amount,
             unit = AreaUnit(forceUnit = this.unit,
                             pressureUnit = pressure.unit))

    open operator fun div(area: Area) =
        Pressure(amount = this.amount / area.amount,
                 unit = PressureUnit(forceUnit = this.unit,
                                     areaUnit = area.unit))

    // endregion
}